From 4f05d57150e038cdc6528518f27c96a4581b4766 Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@cavium.com>
Date: Tue, 5 Mar 2019 14:47:45 -0800
Subject: [PATCH 049/386] mmc: cavium: correct clock divisor

Rounding of the mmc clock divisor was incorrect,
allowing overclocking. We still keep clk_lo/_hi identical,
using (n, n+1) for intermediate frequencies is unproven.

Now explicitly check for overclocking.

Adjusting system clock, from which mmc clock is derived,
may allow higher frequencies when the next-smaller
divisor would take mmc clock past device maximum

Change-Id: I464fe91afa94a202e9db2682f52f29dfe1fa9aa5
Signed-off-by: Peter Swain <pswain@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/5239
Reviewed-by: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/mmc/host/cavium.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index 5fee4d2624e9..a05c39619ac5 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -1305,8 +1305,17 @@ static void cvm_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 		clock = max_f;
 	slot->clock = clock;
 
-	if (clock)
-		clk_period = (host->sys_freq + clock - 1) / (2 * clock);
+	if (clock) {
+		clk_period = host->sys_freq / (2 * clock);
+		/* check to not exceed requested speed */
+		while (1) {
+			int hz = host->sys_freq / (2 * clk_period);
+
+			if (hz <= clock)
+				break;
+			clk_period++;
+		}
+	}
 
 	emm_switch =
 		     FIELD_PREP(MIO_EMM_SWITCH_BUS_WIDTH, bus_width) |
-- 
2.17.1

